### ISSUE TITLE
Watcher emits create event for quickly deleting and recreating a file

### ISSUE THREAD

**Issue by bpasero** [parcel-bundler/watcher#72]
VS Code itself does some event coalescing (see [watcher.ts](https://github.com/microsoft/vscode/blob/a4342f3addc29f405a5cb4aab359b0b98223411e/src/vs/platform/files/node/watcher/watcher.ts#L64)). I think these rules can be pushed down to the watcher itself so that everyone benefits from this.

I would have expected the event coalescing to emit a change event for this case where a file is quickly deleted and then recreated again:

```js
(async function () {
    let subscription = await watcher.subscribe('<path to folder>', (err, events) => {
        console.log(err, events);
    }).catch(err => {
        console.error(err);
    }).then(() => {
        setTimeout(() => {
            fs.unlinkSync('<path to existing file>');
            setTimeout(() => {
                fs.writeFileSync('<path to existing file>', 'Hello World');
            }, 1);
        }, 1);
    });
})();
```

From my testing on macOS and Windows, I get a single create event.

The scenario where this is interesting is applications that do "atomic" writes to files, meaning they create a tmp file, write the contents there and then move it over to the target file to have a single file operation for the write.

**Comment by devongovett** [comment]
Same reply as #71.

**Comment by bpasero** [comment]
Does that apply even to Windows? My tests resulted in the same on both platforms.

This might be a bit harder for VS Code to digest in case of atomic writes. But I need to test with our previous watchers if they behaved any different...

**Comment by devongovett** [comment]
Hmm I think the logic is here:

https://github.com/parcel-bundler/watcher/blob/05a4b23869c6dea015e1cb0bc98f75bdac4a28fb/src/Event.hh#L20

Perhaps we should have a case for both created and deleted that emits "updated" instead? It would still not happen all the time on macOS but should be a bit better elsewhere.

**Comment by bpasero** [comment]
I am not sure this kind of event normalization should happen in the watcher: in the end what happens on disk is a delete followed by a create, so if anyone wants to normalize that to an "updated" event, they should do it.

But if the watcher itself does not emit these 2 events, a client cannot do this.

I wonder if your structure of known files could help here (`State`)?

**Comment by bpasero** [comment]
This issue is now being reported as https://github.com/microsoft/vscode/issues/136460 for VS Code where it looks like the `git` process deletes and writes the `.git/HEAD` file very quickly. Previously extensions would listen to change events on this file to react and now have to change to listen to create events.

I was able to verify that our previous watchers reported the create+delete events and we did fold the event into a single change event.
